server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -391,6 +391,12 @@ private synchronized DocumentMapper internalMerge(DocumentMapper mapper, MergeRe
         }
         checkIndexSortCompatibility(indexSettings.getIndexSortConfig(), hasNested);
 
+        for (MetadataFieldMapper metadataFieldMapper : metadataMappers) {
```

Comment:
We could move this logic into `MapperMergeValidator#validateFieldReferences` to keep all the cross-field validation in one place.